plugins/ingest/src/main/java/org/elasticsearch/ingest/Data.java

```diff
@@ -47,67 +48,95 @@ public Data(Data other) {
+    public <T> T getProperty(String path, Class<T> clazz) {
+        Object property = get(path);
+        if (property == null) {
+            return null;
+        }
+        if (clazz.isInstance(property)) {
+            return clazz.cast(property);
+        }
+        throw new IllegalArgumentException("field [" + path + "] of type [" + property.getClass().getName() + "] cannot be cast to [" + clazz.getName() + "]");
     }
 
+    /**
+     * Checks whether the document contains a value for the provided path
+     * @param path The path within the document in dot-notation
+     * @return true if the document contains the property, false otherwise
+     */
     public boolean containsProperty(String path) {
-        boolean containsProperty = false;
-        String[] pathElements = Strings.splitStringToArray(path, '.');
-        if (pathElements.length == 0) {
-            return false;
-        }
-
-        Map<String, Object> inner = document;
+        return get(path) != null;
+    }
 
-        for (int i = 0; i < pathElements.length; i++) {
-            if (inner == null) {
-                containsProperty = false;
-                break;
-            }
-            if (i == pathElements.length - 1) {
-                containsProperty = inner.containsKey(pathElements[i]);
-                break;
-            }
+    private Object get(String path) {
+        if (path == null || path.length() == 0) {
+            return null;
+        }
+        String[] pathElements = Strings.splitStringToArray(path, '.');
+        assert pathElements.length > 0;
 
-            Object obj = inner.get(pathElements[i]);
+        Map<String, Object> innerMap = document;
+        for (int i = 0; i < pathElements.length - 1; i++) {
+            Object obj = innerMap.get(pathElements[i]);
             if (obj instanceof Map) {
-                inner = (Map<String, Object>) obj;
+                @SuppressWarnings("unchecked")
+                Map<String, Object> stringObjectMap = (Map<String, Object>) obj;
+                innerMap = stringObjectMap;
             } else {
-                inner = null;
+                return null;
             }
         }
 
-        return containsProperty;
+        String leafKey = pathElements[pathElements.length - 1];
+        return innerMap.get(leafKey);
     }
 
     /**
-     * add `value` to path in document. If path does not exist,
-     * nested hashmaps will be put in as parent key values until
+     * Adds the provided value to the provided path in the document.
+     * If path does not exist, nested maps will be put in as parent key values until
      * leaf key name in path is reached.
-     *
      * @param path The path within the document in dot-notation
      * @param value The value to put in for the path key
      */
     public void addField(String path, Object value) {
+        if (path == null || path.length() == 0) {
+            throw new IllegalArgumentException("cannot add null or empty field");
```

Comment:
why not?

what if a user is intentionally allowing null to enable more flexible parsing during mutation. For example, Grok may be extracting values out of fields, and the document either has that property or it doesn't, but the field is ok to be null, as a placeholder.